### DISCUSSION TITLE
Hosting/Starting up multiple levels

### DISCUSSION THREAD

**Discussion by ghost** [o3de/o3de#8604] [General]
As a dev I'm asking for a guidance/documentation for being able to start instances on different levels in a persistent multiplayer world environment.

When the server launcher starts it asks for 1 single level instead of multiple levels. That is fine for a "map" from a first-person shooter game perspective, but how do we design architecturally a system to fufill the needs of a persistent world environt such as an MMO, online adventure, where players travel to different areas of this world.

Please advice whether this requires another approach in the core engine, networking, or both.

I've been in touch with Yuyi and Seapip and we more than willing to collaborate on this matter and provide assistance with documentating this approach for future members of the community who might encounter the same range of questions.

We are in need to guidance/technical advice if this is possible, or code changes on the server-portion of the engine to be able to spawn multiple, simultaneously levels that co-exist, regardless if clients or connected to that environment or not.

To summarize our request,

1. How do we initiate more than 1 levels on the same server on server-startup?
2. Where in the code could/should we begin initiating multiple instances of levels, accessible to clients?

**Comment by lmbr-pip** [comment]
It might be nice to get some more details about what you are trying to do. What requirements do you have? How large a world area are you trying to build? What player population per server are you aiming for?

Mostly trying to understand why the request for a server to host multiple levels comes from. Is there a reason not to have a large map where terrain or disjoint areas to create the separation you need?

**Comment by lmbr-pip** [comment]
@kberg-amzn - Do you have any advice here? Will ensure this is tracked as potential item for next SIG-network meeting

**Comment by ghost** [comment]
> It might be nice to get some more details about what you are trying to do. What requirements do you have? How large a world area are you trying to build? What player population per server are you aiming for?
> 
> Mostly trying to understand why the request for a server to host multiple levels comes from. Is there a reason not to have a large map where terrain or disjoint areas to create the separation you need?

Hello, 

I just want to clarify whether or not this is a probable design by O3DE or not. It is an architectural question where load balancing and instantiation comes to mind.

It is hard to point at numbers at this early stage in development, it is very much like predicting the weather. Our view would be around 200 simultaneous accesses by individual clients per level, before considering to "lift" the level to a new instance.

Is currently the server launcher instance designed to be hosted in single server environment? If yes, would that mean that you host only 1 level per launched server instance?

That question is also important when it comes to infrastructure and trying to plan for license costs on OS and other software.

Thanks

**Comment by kberg0** [comment]
If I understand the request, they want distinct regions and zones that are each saved as separate prefabs/levels/spawnables. This might be more a question for gameobjects and sig-core, and honestly it's not really a networking specific question; but my first instinct would be to author each of those zones as prefabs, have an essentially empty root map, and then spawn in the various zones required at runtime. This means the server could load and unload content as players move throughout the various segments of the world.

**Comment by ghost** [comment]
> If I understand the request, they want distinct regions and zones that are each saved as separate prefabs/levels/spawnables. This might be more a question for gameobjects and sig-core, and honestly it's not really a networking specific question; but my first instinct would be to author each of those zones as prefabs, have an essentially empty root map, and then spawn in the various zones required at runtime. This means the server could load and unload content as players move throughout the various segments of the world.

That is however memory and processing consuming workflow and not really recommended for an MMO where the world is available at all times, whether players are present or not. I think I mentioned MMO  in the 2nd paragraph of my original post.

**Comment by kberg0** [comment]
Again, I would bring it up with sig-core and the game objects team, this is their domain and I am not an expert here, but as of currently the engine does not support multiple spatially overlapping but otherwise non-interactive scenes or maps or levels. If you need multiple zones, then I believe your two choices are either to put all of the different zones into a single space, or to run multiple server processes and migrate players between servers as they transition zones. It's certainly possible to run multiple server processes on a single machine.

**Comment by SelfishOlex** [comment]
Fyi, you can avoid LoadLevel and I just spawn the prefab of the level using C++ Spawning Prefab API.

**Comment by lmbr-pip** [comment]
@AMZN-Olex thinks you can spawn a prefab to do this, as a prefab can be a level. Open questions around splitting simulation and other game engine components.

* Does this all need to happen in the same process? Can you split zones across processes?

**Comment by lmbr-pip** [comment]
Migrated to a Discussion/Discord thread to continue discussion.

**Comment by ghost** [comment]
@AMZN-Olex @lmbr-pip @kberg-amzn could you please tag me when this moves to a thread?

Additionally, how do you distinguish code that runs by the client or server? When can we tell the code is executed server-side vs. client-side separately?

**Comment by lmbr-pip** [comment]
There are a couple of mechanisms for client/server, today:
* You can do somethings in CMake to separate code, but things will be cleaner once: https://github.com/o3de/sig-network/blob/main/rfcs/rfc-net-20220118-1-clientserver.md is delivered later this year. 

But you can do things like this at the top of a CMakeList.txt file where the file contains your server components, so the code is only included in the server target.
```
if (NOT PAL_TRAIT_BUILD_SERVER_SUPPORTED)
    return()
endif()
```

* You can use the MultiplayerAgentType to find out what you are running as (if using Multiplayer): https://github.com/o3de/o3de/blob/7f5b87085c464c2523e746e13ea11bb26a700853/Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayer.h#L28 if you have generic code that runs in both places.

It does depend on when you want to make this decision.

**Reply by ghost** [reply]
> There are a couple of mechanisms for client/server, today:
> 
> * You can do somethings in CMake to separate code, but things will be cleaner once: https://github.com/o3de/sig-network/blob/main/rfcs/rfc-net-20220118-1-clientserver.md is delivered later this year.
> 
> But you can do things like this at the top of a CMakeList.txt file where the file contains your server components, so the code is only included in the server target.
> 
> ```
> if (NOT PAL_TRAIT_BUILD_SERVER_SUPPORTED)
>     return()
> endif()
> ```
> 
> * You can use the MultiplayerAgentType to find out what you are running as (if using Multiplayer): https://github.com/o3de/o3de/blob/7f5b87085c464c2523e746e13ea11bb26a700853/Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayer.h#L28
>    if you have generic code that runs in both places.
> 
> It does depend on when you want to make this decision.

Thank you @lmbr-pip I'll try it right away!
